content/en/blog/2022-03-21-tools-to-manage-helm-declaratively.md

```diff
@@ -0,0 +1,86 @@
+
+Kubernetes provides a means to do both [declarative and imperative management of resources](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/). As the Kubernetes community tends to prefer declarative management, when possible, the rest of this post is going to focus on declarative tools you can use with Helm.
+
+## Tools
+
+The Kubernetes ecosystem has produced numerous projects of various styles to help you declaratively manage your Helm releases. To illustrate the options we will look at sister projects to Helm in the Cloud Native Computing Foundation (CNCF) and some more general open source projects. You can find more options in the [CNCF Landscape](https://landscape.cncf.io/).
+
+### CNCF Projects
+
+The scope of this section is limited to [graduated and incubating](https://www.cncf.io/projects/) CNCF projects. There are over 100 CNCF projects and many of them are [sandbox projects](https://www.cncf.io/sandbox-projects/). You can learn more about the differences between types of projects in the [maturity level explanation](https://www.cncf.io/projects/#maturity-levels). The following projects are worth looking at:
+
+* [Flux Helm Controller](https://fluxcd.io/docs/components/helm/) - [Flux](https://fluxcd.io/) is a collection of projects that enable GitOps. One of the components provides a GitOps method to manage Helm releases.
+* [Argo CD](https://github.com/argoproj/argo-cd) - The [Argo](https://argoproj.github.io/) project defines itself as providing "Open source tools for Kubernetes to run workflows, manage clusters, and do GitOps right." Argo CD is focused on declarative continuous delivery and has the ability to work with Helm charts.
+
+### Other Projects
+
+There are many projects beyond the CNCF projects you can use to help you manage your Helm releases. The following set is an example and not exhaustive.
+
+* [Helmfile](https://github.com/roboll/helmfile) - A declarative spec for deploying Helm charts.
+* [Captain](https://github.com/alauda/captain) - A Helm controller.
+* [Terraform Helm provider](https://github.com/hashicorp/terraform-provider-helm) - Enables you to manage Helm charts through Terraform.
+* [Orkestra](https://azure.github.io/orkestra/) - Built on other tools in this list, Orkestra adds a robust dependency graph for a related group of Helm releases and their subcharts, as well as a reverse DAG for specifying dependency requirements for rollbacks.
+* [Fleet](https://github.com/rancher/fleet) - A GitOps tool chain that works with Kubernetes manifests, Helm charts, and Kustomize.
+
+### High-Level Tool Comparison
+
+There are some differences between the tools we've looked at so far. The following table provides some insight into their differences. This is not exhaustive and you should evaluate any tools you use yourself.
+
+| | Retains Helm release info | Supports Helm hooks | OCI support | Does not require Helm binary |
```

Comment:
@yurrriq I see your reply to Matt in the [HackMD doc](https://hackmd.io/_hSNKSDhTPihfCtyT8nCGA?view) comment for the removed Helm SDK column. 

@mattfarina I do agree with you that whether a declarative tool uses the Helm SDK is less relevant to end-users, but I think @yurrriq makes a good point that end-users may want to know which tools require the `helm` binary (and if so which version) to be on the same system running the tool, especially when running in a container. Copying HackMD convo here for context. @mattfarina do you agree with his reasoning?

> @mattfarina
> The target audience for this post is someone looking for tooling and not tool authors. That’s why the Helm SDK column was dropped from the table. Not sure why a consumer is going to care. Am I wrong? If so, why? Trying to focus.
>> @scottrigby 
>> This is a good question. In general I'm trying to raise awareness of the value of Helm's SDK to end users, even though – apart from developers working with it directly – most people's only interaction with Helm's SDK is through other tools that import and use the it. I agree this post isn't focusing on the SDK benefits though, so I'm on the fence about whether that detail is interesting/useful for folks as a table column. Maybe it is? Would love other thoughts on this 🤔
>>> @mattfarina
>>> I may have mentioned this before but I'm a fan of the Jobs to be Done theory. End users have a job they need to be done. How that sausage is made... and even what the sausage is... they don't really care about as long as the job gets done. More details at https://www.christenseninstitute.org/jobs-to-be-done 
>>>
>>> The SDK is middleware. Why would an end user who has some job (i.e. my workload is running in Kubernetes) care about the middleware? Does someone use the SDK? Does someone do something else?
>>>> @yurrriq 
>>>> I think the distinction is important, even to an end user. Some examples:
>>>> * As an end user, I might be expecting a feature from a newer version of Helm, which I have on my PATH, to be supported by the tool. And maybe it's not.
>>>> * If I want to include a tool in a Docker image, I need to know whether I need to include a Helm binary as well.

Reply:
Thanks for bring the conversation over here @scottrigby 

I think it's a good point to know what features you get with Helm and one of these. Does it use the Helm SDK? If so, at what version and with what features? For example, fleet uses the Helm SDK at a version prior to OCI registries. If it uses the Helm binary you need that in your path.

This is why I noted, repeatedly, that people need to evaluate the projects for themselves. For example,
> This is not exhaustive and you should evaluate any tools you use yourself.

There is enough room in the table to add a column to state of the Helm binary is needed or not. So, I am happy that Scott added that.